operator/pkg/gateway-api/gateway.go

```
@@ -139,85 +201,214 @@ func (r *gatewayReconciler) enqueueRequestForOwningGatewayClass() handler.EventH
+		if err := r.Client.List(ctx, gwList, &client.ListOptions{
+			FieldSelector: fields.OneTermEqualSelector(implementationGatewayIndex, "cilium"),
+		}); err != nil {
+			scopedLog.ErrorContext(ctx, "Failed to get Cilium Gateways", logfields.Error, err)
+			return []reconcile.Request{}
+		}
+
+		// Build a set of all Cilium Gateway full names.
+		// This makes sure we only add a reconcile.Request once for each Gateway.
+		allCiliumGatewaysSet := make(map[string]struct{})
+		for _, gw := range gwList.Items {
+			gwFullName := types.NamespacedName{
+				Name:      gw.GetName(),
+				Namespace: gw.GetNamespace(),
+			}
+			allCiliumGatewaysSet[gwFullName.String()] = struct{}{}
+		}
+
+		// iterate through the HTTPRoutes, return a reconcile.Request for each Gateways that is relevant.
+		for _, hr := range hrList.Items {
+			for _, parent := range hr.Spec.ParentRefs {
+				if !helpers.IsGateway(parent) {
+					continue
+				}
+				parentFullName := types.NamespacedName{
+					Name:      string(parent.Name),
+					Namespace: helpers.NamespaceDerefOr(parent.Namespace, hr.Namespace),
+				}
+				if _, found := allCiliumGatewaysSet[parentFullName.String()]; found {
+					reconcileRequests[reconcile.Request{NamespacedName: parentFullName}] = struct{}{}
+				}
+			}
+		}
+
+		// return the keys of the set.
+		return slices.Collect(maps.Keys(reconcileRequests))
 	})
 }
 
-func getReconcileRequestsForRoute(ctx context.Context, c client.Client, object metav1.Object, route gatewayv1.CommonRouteSpec, logger *slog.Logger) []reconcile.Request {
+func getGatewayReconcileRequestsForRoute(ctx context.Context, c client.Client, object metav1.Object, route gatewayv1.CommonRouteSpec, logger *slog.Logger) []reconcile.Request {
 	var reqs []reconcile.Request
 
 	scopedLog := logger.With(
-		logfields.Controller, gateway,
-		logfields.Resource, types.NamespacedName{
-			Namespace: object.GetNamespace(),
-			Name:      object.GetName(),
-		},
+		logfields.Resource, object.GetName(),
```

It looks like the following code silently skips `HTTPRoute`s that don't reference a `Gateway`.

```
		if !helpers.IsGateway(parent) {
			continue
		}
```

Looking at the Gateway API spec (`parentRefs`: https://gateway-api.sigs.k8s.io/reference/spec/#httproutespec)  it seems that the same `HTTPRoute` could reference a `Gateway` (Gateway API) and a K8s `Service` (GAMMA) - as long as the refs are distinct. Is this correct or am I interpreting this wrong?

If this use-case is allowed by the spec (whether it makes sense or not is another question), (and we would loosen above restriction) - we might face the potential risk of two reconcilers (Gateway & GAMMA) fighting over (the status) of the same `HTTPRoute`?

If this use-case isn't allowed by the spec, i think we should surface this issue to the user via status.

---

It is absolutely possible that the same HTTPRoute could reference both a Gateway and a Service - but that's why the HTTPRoute status is namespaced by the parent. Some of the tests I added a while ago are actually about making sure that the two reconcilers don't clobber each other's status.

So yes, it is allowed by the spec, but the problem is already handled.

Thanks for the callout though!

---

ah yes, nice.

I completely missed this because my local reproducer failed to accept the GAMMA HTTPRoute because we don't support the parentRef pointing to a K8s Service using the group `core` (we validate that it needs to be an empty string `""`). Now that i fixed this i see the two separate status reported by both parents. Great.

Will open a PR to fix this.

---

Opened PR: https://github.com/cilium/cilium/pull/41268 (shouldn't conflict with this PR)

---

hmm, I read over this, and to clarify: we have the two reconcilers, so each can handle _either_ a Gateway ParentRef _or_ a GAMMA ParentRef. The changes I made a while ago were to ensure that they don't stomp on each other's status, and there are unit tests for that from both directions. I'm also going to take those unit tests upstream and turn them into conformance tests (probably for v1.5, since code freeze for v1.4 is Monday).